mcxtrace-antlr partly uses code intended for McStas

It seems the `mcxtrace-antlr` is partially configuring with use of McStas rather than McXtrace, since the particle struct from `mccode-r.h(.in)` includes `vx/y/z` and not `kx/y/z`

@g5t no real urgency here, just thought you'd like to know.

```
(base) willend@wshome:~/tmp/McXtrace$ mcxtrace-antlr -I$MCXTRACE template_simple.instr 
Downloading file 'common/lib/share/mccode-r.h.in' from 'https://github.com/McStasMcXtrace/McCode/raw/v3.5.22/common/lib/share/mccode-r.h.in' to '/home/willend/.cache/mccode_antlr/libc/v3.5.22'.
No initialization present?

-----------------------------------------------------------

Generating single GPU kernel or single CPU section layout:

-----------------------------------------------------------

Generating GPU/CPU -DFUNNEL layout:
-> GPU kernel from component origin

-----------------------------------------------------------
(base) willend@wshome:~/tmp/McXtrace$ gcc -lm template_simple.c 
template_simple.c: In function 'particle_getvar':
template_simple.c:109:40: error: '_class_particle' {aka 'struct _struct_particle'} has no member named 'vx'; did you mean 'x'?
  109 |       if(!str_comp("vx",name)){rval=p->vx;s=0;}
      |                                        ^~
      |                                        x
template_simple.c:110:40: error: '_class_particle' {aka 'struct _struct_particle'} has no member named 'vy'; did you mean 'y'?
  110 |       if(!str_comp("vy",name)){rval=p->vy;s=0;}
      |                                        ^~
      |                                        y
template_simple.c:111:40: error: '_class_particle' {aka 'struct _struct_particle'} has no member named 'vz'; did you mean 'z'?
  111 |       if(!str_comp("vz",name)){rval=p->vz;s=0;}
      |                                        ^~
      |                                        z
template_simple.c:112:40: error: '_class_particle' {aka 'struct _struct_particle'} has no member named 'sx'; did you mean 'x'?
  112 |       if(!str_comp("sx",name)){rval=p->sx;s=0;}
      |                                        ^~
      |                                        x
template_simple.c:113:40: error: '_class_particle' {aka 'struct _struct_particle'} has no member named 'sy'; did you mean 'y'?
  113 |       if(!str_comp("sy",name)){rval=p->sy;s=0;}
      |                                        ^~
      |                                        y
template_simple.c:114:40: error: '_class_particle' {aka 'struct _struct_particle'} has no member named 'sz'; did you mean 'z'?
  114 |       if(!str_comp("sz",name)){rval=p->sz;s=0;}
      |                                        ^~
      |                                        z
template_simple.c: In function 'particle_getvar_void':
template_simple.c:142:49: error: '_class_particle' {aka 'struct _struct_particle'} has no member named 'vx'; did you mean 'x'?
  142 |       if(!str_comp("vx",name)){rval=(void*)&(p->vx);s=0;}
      |                                                 ^~
      |                                                 x
template_simple.c:143:49: error: '_class_particle' {aka 'struct _struct_particle'} has no member named 'vy'; did you mean 'y'?
  143 |       if(!str_comp("vy",name)){rval=(void*)&(p->vy);s=0;}
      |                                                 ^~
      |                                                 y
template_simple.c:144:49: error: '_class_particle' {aka 'struct _struct_particle'} has no member named 'vz'; did you mean 'z'?
  144 |       if(!str_comp("vz",name)){rval=(void*)&(p->vz);s=0;}
      |                                                 ^~
      |                                                 z
template_simple.c:145:49: error: '_class_particle' {aka 'struct _struct_particle'} has no member named 'sx'; did you mean 'x'?
  145 |       if(!str_comp("sx",name)){rval=(void*)&(p->sx);s=0;}
      |                                                 ^~
      |                                                 x
template_simple.c:146:49: error: '_class_particle' {aka 'struct _struct_particle'} has no member named 'sy'; did you mean 'y'?
  146 |       if(!str_comp("sy",name)){rval=(void*)&(p->sy);s=0;}
      |                                                 ^~
      |                                                 y
template_simple.c:147:49: error: '_class_particle' {aka 'struct _struct_particle'} has no member named 'sz'; did you mean 'z'?
  147 |       if(!str_comp("sz",name)){rval=(void*)&(p->sz);s=0;}
      |                                                 ^~
      |                                                 z
template_simple.c: In function 'particle_setvar_void':
template_simple.c:166:44: error: '_class_particle' {aka 'struct _struct_particle'} has no member named 'vx'; did you mean 'x'?
  166 |       if(!str_comp("vx",name)){memcpy(&(p->vx), value, sizeof(double)); rval=0;}
      |                                            ^~
      |                                            x
template_simple.c:167:44: error: '_class_particle' {aka 'struct _struct_particle'} has no member named 'vy'; did you mean 'y'?
  167 |       if(!str_comp("vy",name)){memcpy(&(p->vy), value, sizeof(double)); rval=0;}
      |                                            ^~
      |                                            y
template_simple.c:168:44: error: '_class_particle' {aka 'struct _struct_particle'} has no member named 'vz'; did you mean 'z'?
  168 |       if(!str_comp("vz",name)){memcpy(&(p->vz), value, sizeof(double)); rval=0;}
      |                                            ^~
      |                                            z
template_simple.c:169:44: error: '_class_particle' {aka 'struct _struct_particle'} has no member named 'sx'; did you mean 'x'?
  169 |       if(!str_comp("sx",name)){memcpy(&(p->sx), value, sizeof(double)); rval=0;}
      |                                            ^~
      |                                            x
template_simple.c:170:44: error: '_class_particle' {aka 'struct _struct_particle'} has no member named 'sy'; did you mean 'y'?
  170 |       if(!str_comp("sy",name)){memcpy(&(p->sy), value, sizeof(double)); rval=0;}
      |                                            ^~
      |                                            y
template_simple.c:171:44: error: '_class_particle' {aka 'struct _struct_particle'} has no member named 'sz'; did you mean 'z'?
  171 |       if(!str_comp("sz",name)){memcpy(&(p->sz), value, sizeof(double)); rval=0;}
      |                                            ^~
      |                                            z
template_simple.c: In function 'particle_restore':
template_simple.c:195:10: error: '_class_particle' {aka 'struct _struct_particle'} has no member named 'vx'; did you mean 'x'?
  195 |       p->vx = p0->vx; p->vy = p0->vy; p->vz = p0->vz;
      |          ^~
      |          x
template_simple.c:195:19: error: '_class_particle' {aka 'struct _struct_particle'} has no member named 'vx'; did you mean 'x'?
  195 |       p->vx = p0->vx; p->vy = p0->vy; p->vz = p0->vz;
      |                   ^~
      |                   x
template_simple.c:195:26: error: '_class_particle' {aka 'struct _struct_particle'} has no member named 'vy'; did you mean 'y'?
  195 |       p->vx = p0->vx; p->vy = p0->vy; p->vz = p0->vz;
      |                          ^~
      |                          y
template_simple.c:195:35: error: '_class_particle' {aka 'struct _struct_particle'} has no member named 'vy'; did you mean 'y'?
  195 |       p->vx = p0->vx; p->vy = p0->vy; p->vz = p0->vz;
      |                                   ^~
      |                                   y
template_simple.c:195:42: error: '_class_particle' {aka 'struct _struct_particle'} has no member named 'vz'; did you mean 'z'?
  195 |       p->vx = p0->vx; p->vy = p0->vy; p->vz = p0->vz;
      |                                          ^~
      |                                          z
template_simple.c:195:51: error: '_class_particle' {aka 'struct _struct_particle'} has no member named 'vz'; did you mean 'z'?
  195 |       p->vx = p0->vx; p->vy = p0->vy; p->vz = p0->vz;
      |                                                   ^~
      |                                                   z
template_simple.c:196:10: error: '_class_particle' {aka 'struct _struct_particle'} has no member named 'sx'; did you mean 'x'?
  196 |       p->sx = p0->sx; p->sy = p0->sy; p->sz = p0->sz;
      |          ^~
      |          x
template_simple.c:196:19: error: '_class_particle' {aka 'struct _struct_particle'} has no member named 'sx'; did you mean 'x'?
  196 |       p->sx = p0->sx; p->sy = p0->sy; p->sz = p0->sz;
      |                   ^~
      |                   x
template_simple.c:196:26: error: '_class_particle' {aka 'struct _struct_particle'} has no member named 'sy'; did you mean 'y'?
  196 |       p->sx = p0->sx; p->sy = p0->sy; p->sz = p0->sz;
      |                          ^~
      |                          y
template_simple.c:196:35: error: '_class_particle' {aka 'struct _struct_particle'} has no member named 'sy'; did you mean 'y'?
  196 |       p->sx = p0->sx; p->sy = p0->sy; p->sz = p0->sz;
      |                                   ^~
      |                                   y
template_simple.c:196:42: error: '_class_particle' {aka 'struct _struct_particle'} has no member named 'sz'; did you mean 'z'?
  196 |       p->sx = p0->sx; p->sy = p0->sy; p->sz = p0->sz;
      |                                          ^~
      |                                          z
template_simple.c:196:51: error: '_class_particle' {aka 'struct _struct_particle'} has no member named 'sz'; did you mean 'z'?
  196 |       p->sx = p0->sx; p->sy = p0->sy; p->sz = p0->sz;
      |                                                   ^~
      |                                                   z
template_simple.c: At top level:
template_simple.c:6469:17: error: conflicting types for 'mcsetstate'; have '_class_particle(double,  double,  double,  double,  double,  double,  double,  double,  double,  double,  double,  double,  int,  void *, int)' {aka 'struct _struct_particle(double,  double,  double,  double,  double,  double,  double,  double,  double,  double,  double,  double,  int,  void *, int)'}
 6469 | _class_particle mcsetstate(double x, double y, double z, double kx, double ky, double kz,
      |                 ^~~~~~~~~~
template_simple.c:78:21: note: previous declaration of 'mcsetstate' with type '_class_particle(double,  double,  double,  double,  double,  double,  double,  double,  double,  double,  double,  int,  void *, int)' {aka 'struct _struct_particle(double,  double,  double,  double,  double,  double,  double,  double,  double,  double,  double,  int,  void *, int)'}
   78 |     _class_particle mcsetstate(double x, double y, double z, double vx, double vy, double vz,
```

--- a/src/mccode_antlr/reader/registry.py
+++ b/src/mccode_antlr/reader/registry.py
@@ -357,7 +357,7 @@ class InMemoryRegistry(Registry):
         self.name = name
         self.root = '/proc/memory/'  # Something pathlike is needed?
         self.version = mccode_antlr_version()
-        self.components = {k: v for k, v in components.items()}
+        self.components = {k if k.lower().endswith('.comp') else f'{k}.comp': v for k, v in components.items()}
         self.priority = priority
 
     def add(self, name: str, definition: str):

--- a/src/mccode_antlr/translators/c_header.py
+++ b/src/mccode_antlr/translators/c_header.py
@@ -12,6 +12,10 @@ def header_pre_runtime(
 ):
     from datetime import datetime
     from mccode_antlr import version
+    from .c_particle import (
+        accessible_struct_members, restorable_struct_members,
+        setstate_signature_members, getstate_signature_members, setstate_signature_call
+    )
 
     def jump_line(instance, jump):
         return f'long Jump_{instance.name}_{jump.target}; /* the JUMP connection <from>_<to> */'
@@ -36,23 +40,34 @@ def header_pre_runtime(
     uservar_string = '//user variables and comp - injections:\n' if len(uservars) else ''
     uservar_string += '\n'.join([f'  {x};' for x in uservars])
 
+    members = list(accessible_struct_members(is_mcstas)) + [x.name for x in uservars]
     # Shouldn't we exclude array-valued names here?
     getvar = '\n'.join(
-        [f'  if(!str_comp("{x.name}",name)){{rval=*((double*)(&(p->{x.name})));s=0;}}' for x in uservars])
+        [f'  if(!str_comp("{x}",name)){{rval=*((double*)(&(p->{x})));s=0;}}' for x in members]
+    )
 
     # Array valued names seem OK here, since a user can type-cast correctly
     getvar_void = '\n'.join(
-        [f'  if (!str_comp("{x.name}", name)){{rval=(void * ) & (p->{x.name});s=0;}}' for x in uservars])
+        [f'  if (!str_comp("{x}", name)){{rval=(void * ) & (p->{x});s=0;}}' for x in members]
+    )
 
     # For non-array values we can safely use memcpy (probably)
+    members_types = ([(x, y) for x, y in accessible_struct_members(is_mcstas).items()]
+                     + [(x.name, x.dtype) for x in uservars if not x.is_array and not x.is_pointer]
+                     )
     setvar_void = '\n'.join(
-        [f'  if(!str_comp("{x.name}",name)){{memcpy(&(p->{x.name}), value, sizeof({x.dtype})); rval=0;}}'
-         for x in uservars if not x.is_array and not x.is_pointer])
+        [f'  if(!str_comp("{x}",name)){{memcpy(&(p->{x}), value, sizeof({y})); rval=0;}}'
+         for x, y in members_types]
+    )
 
     # an array -- need to know how many elements we're copying
     setvar_void_array = '\n'.join(
         [f'  if(!str_comp("{x.name}",name)){{memcpy(&(p->{x.name}), value, elements * sizeof({x.dtype})); rval=0;}}'
          for x in uservars if x.is_array or x.is_pointer])
+
+    restore = '\n'.join(
+        [f'  p->{x} = p0->{x};' for x in restorable_struct_members(is_mcstas)]
+    )
 
     getuservar_byid = '\n'.join(
         [f'  case {i}: {{rval=*( (double *)(&(p->{x.name})) ); s=0; break;}}' for i, x in enumerate(uservars)])
@@ -65,6 +80,10 @@ def header_pre_runtime(
             print('  --> and may need specific per-particle initialization through an EXTEND block!\n')
         else:
             uservar_init += f'\np->{x.name}={0 if x.init is None else x.init};'
+
+    mcsetstate = ', '.join(f'{t} {n}' for n, t in setstate_signature_members(is_mcstas).items())
+    mcgetstate = ', '.join(f'{t} {n}' for n, t in getstate_signature_members(is_mcstas).items())
+    mcinitstate = ', '.join(setstate_signature_call(is_mcstas))
 
     contents = dedent(f"""/* Automatically generated file. Do not edit.
      * Format:     ANSI C source code
@@ -142,8 +161,9 @@ def header_pre_runtime(
     #pragma acc routine
     _class_particle mcgenstate(void);
     #pragma acc routine
-    _class_particle mcsetstate(double x, double y, double z, double vx, double vy, double vz,
-        double t, double sx, double sy, double sz, double p, int mcgravitation, void *mcMagnet, int mcallowbackprop);
+    _class_particle mcsetstate({mcsetstate});
+    #pragma acc routine
+    _class_particle mcgetstate({mcgetstate});
     
     extern int mcgravitation;      /* flag to enable gravitation */
     #pragma acc declare create ( mcgravitation )
@@ -151,7 +171,7 @@ def header_pre_runtime(
     #pragma acc declare create ( mcallowbackprop )
     
     _class_particle mcgenstate(void) {{
-      _class_particle particle = mcsetstate(0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, mcgravitation, NULL, mcallowbackprop);
+      _class_particle particle = mcsetstate({mcinitstate});
       return(particle);
     }}
     /*Generated user variable handlers:*/
@@ -170,20 +190,6 @@ def header_pre_runtime(
     #endif
       int s=1;
       double rval=0;
-      if(!str_comp("x",name)){{rval=p->x;s=0;}}
-      if(!str_comp("y",name)){{rval=p->y;s=0;}}
-      if(!str_comp("z",name)){{rval=p->z;s=0;}}
-      if(!str_comp("vx",name)){{rval=p->vx;s=0;}}
-      if(!str_comp("vy",name)){{rval=p->vy;s=0;}}
-      if(!str_comp("vz",name)){{rval=p->vz;s=0;}}
-      if(!str_comp("sx",name)){{rval=p->sx;s=0;}}
-      if(!str_comp("sy",name)){{rval=p->sy;s=0;}}
-      if(!str_comp("sz",name)){{rval=p->sz;s=0;}}
-      if(!str_comp("t",name)){{rval=p->t;s=0;}}
-      if(!str_comp("p",name)){{rval=p->p;s=0;}}
-      if(!str_comp("_mctmp_a",name)){{rval=p->_mctmp_a;s=0;}}
-      if(!str_comp("_mctmp_b",name)){{rval=p->_mctmp_b;s=0;}}
-      if(!str_comp("_mctmp_c",name)){{rval=p->_mctmp_c;s=0;}}
     {getvar}  
       if (suc!=0x0) {{*suc=s;}}
       return rval;
@@ -202,17 +208,6 @@ def header_pre_runtime(
     #endif
       int s=1;
       void* rval=0;
-      if(!str_comp("x",name)) {{rval=(void*)&(p->x); s=0;}}
-      if(!str_comp("y",name)) {{rval=(void*)&(p->y); s=0;}}
-      if(!str_comp("z",name)) {{rval=(void*)&(p->z); s=0;}}
-      if(!str_comp("vx",name)){{rval=(void*)&(p->vx);s=0;}}
-      if(!str_comp("vy",name)){{rval=(void*)&(p->vy);s=0;}}
-      if(!str_comp("vz",name)){{rval=(void*)&(p->vz);s=0;}}
-      if(!str_comp("sx",name)){{rval=(void*)&(p->sx);s=0;}}
-      if(!str_comp("sy",name)){{rval=(void*)&(p->sy);s=0;}}
-      if(!str_comp("sz",name)){{rval=(void*)&(p->sz);s=0;}}
-      if(!str_comp("t",name)) {{rval=(void*)&(p->t); s=0;}}
-      if(!str_comp("p",name)) {{rval=(void*)&(p->p); s=0;}}
     {getvar_void}
       if (suc!=0x0) {{*suc=s;}}
       return rval;
@@ -225,17 +220,6 @@ def header_pre_runtime(
     #define str_comp strcmp
     #endif
       int rval=1;
-      if(!str_comp("x",name)) {{memcpy(&(p->x),  value, sizeof(double)); rval=0;}}
-      if(!str_comp("y",name)) {{memcpy(&(p->y),  value, sizeof(double)); rval=0;}}
-      if(!str_comp("z",name)) {{memcpy(&(p->z),  value, sizeof(double)); rval=0;}}
-      if(!str_comp("vx",name)){{memcpy(&(p->vx), value, sizeof(double)); rval=0;}}
-      if(!str_comp("vy",name)){{memcpy(&(p->vy), value, sizeof(double)); rval=0;}}
-      if(!str_comp("vz",name)){{memcpy(&(p->vz), value, sizeof(double)); rval=0;}}
-      if(!str_comp("sx",name)){{memcpy(&(p->sx), value, sizeof(double)); rval=0;}}
-      if(!str_comp("sy",name)){{memcpy(&(p->sy), value, sizeof(double)); rval=0;}}
-      if(!str_comp("sz",name)){{memcpy(&(p->sz), value, sizeof(double)); rval=0;}}
-      if(!str_comp("p",name)) {{memcpy(&(p->p),  value, sizeof(double)); rval=0;}}
-      if(!str_comp("t",name)) {{memcpy(&(p->t),  value, sizeof(double)); rval=0;}}
     {setvar_void}  
       return rval;
     }}
@@ -256,10 +240,7 @@ def header_pre_runtime(
     #pragma acc routine
     void particle_restore(_class_particle *p, _class_particle *p0);
     void particle_restore(_class_particle *p, _class_particle *p0) {{
-      p->x  = p0->x;  p->y  = p0->y;  p->z  = p0->z;
-      p->vx = p0->vx; p->vy = p0->vy; p->vz = p0->vz;
-      p->sx = p0->sx; p->sy = p0->sy; p->sz = p0->sz;
-      p->t = p0->t;  p->p  = p0->p;
+      {restore}
       p->_absorbed=0; p->_restore=0;
     }}
     

--- a/src/mccode_antlr/translators/c_particle.py
+++ b/src/mccode_antlr/translators/c_particle.py
@@ -1,0 +1,123 @@
+def xyz_Axyz_Bxyz(is_mcstas, type_str):
+    """Return lists of particle struct members
+
+    Either:
+        (x, y, z), (vx, vy, vz), (sx, sy, sz)
+    or
+        (x, y, z), (kx, ky, kz, phi), (Ex, Ey, Ez)
+    """
+    xyz = ('x', 'y', 'z')
+    ps = ('', 'v', 's') if is_mcstas else ('', 'k', 'E')
+    z, a, b = (tuple((f'{p}{x}', type_str) for x in xyz) for p in ps)
+    if not is_mcstas:
+        a = a + (('phi', type_str),)
+    return z, a, b
+
+
+
+def struct_members(is_mcstas: bool):
+    """Get the name and type of the members of the _class_particle struct
+
+    The members of the struct are not equivalent between McStas and McXtrace
+    """
+    z, a, b = xyz_Axyz_Bxyz(is_mcstas, 'double')
+    if is_mcstas:
+        a = a +  (
+            ('mcgravitation', 'int'),
+            ('mcMagnet', 'void *'),
+            ('allow_backprop', 'int')
+        )
+
+    members = (z + a + b) + (
+        ('_mctmp_a', 'double'),
+        ('_mctmp_b', 'double'),
+        ('_mctmp_c', 'double'),
+        ('randstate', 'unsigned long []'),
+        ('t', 'double'),
+        ('p', 'double'),
+        ('_uid', 'long long'),
+        ('_index', 'long'),
+        ('_absorbed', 'long'),
+        ('_scattered', 'long'),
+        ('_restore', 'long'),
+        ('flag_nocoordschange', 'long'),
+        ('_logic', 'struct particle_logic_struct'),
+        # uservars
+    )
+    return {k: v for k, v in members}
+
+def accessible_struct_members(is_mcstas: bool):
+    """Get the name and type of the accessible members of the _class_particle struct
+
+    The members of the struct are not equivalent between McStas and McXtrace
+
+    Note
+    ----
+    Used in the generated C code by, e.g.,
+        double particle_getvar(...)
+        void particle_getvar_void(...)
+        int particle_setvar_void(...)
+        void particle_restore(...)
+    written-out here within c_header.py
+    """
+    z, a, b = xyz_Axyz_Bxyz(is_mcstas, 'double')
+    members = (z + a + b) + (
+        ('t', 'double'),
+        ('p', 'double'),
+        ('_mctmp_a', 'double'),
+        ('_mctmp_b', 'double'),
+        ('_mctmp_c', 'double'),
+        # uservars
+    )
+    return {k: v for k, v in members}
+
+def restorable_struct_members(is_mcstas: bool):
+    """Get the name and type of the restorable members of the _class_particle struct
+
+    The members of the struct are not equivalent between McStas and McXtrace
+
+    Note
+    ----
+    Used in the generated C code by, e.g.,
+        void particle_restore(...)
+    written-out here within c_header.py
+    """
+    z, a, b = xyz_Axyz_Bxyz(is_mcstas, 'double')
+    members = z + a + b + (('t', 'double'), ('p', 'double'),)
+    return {k: v for k, v in members}
+
+
+def setstate_signature_members(is_mcstas: bool):
+    z, a, b = xyz_Axyz_Bxyz(is_mcstas, 'double')
+    members = (
+        z + a + (('t', 'double'),) + b
+        + (
+            ('p', 'double'),
+            ('mcgravitation', 'int'),
+            ('mcMagnet', 'void *'),
+            ('allow_backprop', 'int')
+        )
+    )
+    return {k: v for k, v in members}
+
+
+def getstate_signature_members(is_mcstas):
+    z, a, b = xyz_Axyz_Bxyz(is_mcstas, 'double *')
+    members = (
+            (('mcparticle', '_class_particle'),)
+            + z + a + (('t', 'double *'),) + b + (('p', 'double *'),)
+    )
+    return {k: v for k, v in members}
+
+
+
+def setstate_signature_call(is_mcstas: bool):
+    values = {
+        'vz' if is_mcstas else 'kz': '1',
+        'p': '1',
+        'mcgravitation': 'mcgravitation',
+        'mcMagnet': 'NULL',
+        'allow_backprop': 'mcallowbackprop',
+    }
+    members = setstate_signature_members(is_mcstas)
+    return [values.get(key, '0') for key in members]

--- a/tests/runtime/compiled.py
+++ b/tests/runtime/compiled.py
@@ -1,5 +1,6 @@
 from __future__ import annotations
 from functools import cache
+from enum import Enum
 import pytest
 
 from mccode_antlr.compiler.check import simple_instr_compiles
@@ -45,13 +46,26 @@ def mcpl_compiled_test(method):
     return method
 
 
-def compile_and_run(instr, parameters, run=True, dump_source=True, target: dict | None = None, config: dict | None = None):
+class Flavor(Enum):
+    MCSTAS=1
+    MCXTRACE=2
+
+
+def compile_and_run(instr,
+                    parameters,
+                    run=True,
+                    dump_source=True,
+                    target: dict | None = None,
+                    config: dict | None = None,
+                    flavor: Flavor = Flavor.MCSTAS):
     from pathlib import Path
     from tempfile import TemporaryDirectory
-    from mccode_antlr.translators.target import MCSTAS_GENERATOR
+    from mccode_antlr.translators.target import MCSTAS_GENERATOR, MCXTRACE_GENERATOR
     from mccode_antlr.run import mccode_compile, mccode_run_compiled
 
-    kwargs = dict(generator=MCSTAS_GENERATOR, target=target, config=config, dump_source=dump_source)
+    generator = MCXTRACE_GENERATOR if flavor == Flavor.MCXTRACE else MCSTAS_GENERATOR
+
+    kwargs = dict(generator=generator, target=target, config=config, dump_source=dump_source)
 
     with TemporaryDirectory() as directory:
         binary, target = mccode_compile(instr, directory, **kwargs)

--- a/tests/runtime/test_mcxtrace.py
+++ b/tests/runtime/test_mcxtrace.py
@@ -1,0 +1,45 @@
+from textwrap import dedent
+from mccode_antlr.loader import parse_mcstas_instr
+from .compiled import compiled_test, compile_and_run, Flavor
+from mccode_antlr.reader.registry import InMemoryRegistry
+
+in_memory = InMemoryRegistry(
+    "test_components",
+    trace_counter=dedent("""DEFINE COMPONENT trace_counter
+    SETTING PARAMETERS (int n)
+    NOACC
+    TRACE
+    %{
+      for (int i= 0; i < n; i++) {
+        printf("%d\\n", i);
+      }
+    %}
+    END
+    """),
+    useless=dedent("""DEFINE COMPONENT useless
+    SETTING PARAMETERS (int n)
+    TRACE 
+    %{
+    %}
+    END
+    """),
+)
+
+
+@compiled_test
+def test_simplest_mcstas():
+    instr = parse_mcstas_instr(dedent("""
+    define instrument test_simplest_mcstas(dummy=0.)
+    trace component origin = useless() at (0, 0, 0) absolute end
+    """), registries=[in_memory])
+    compile_and_run(instr, '-n 1 dummy=2', flavor=Flavor.MCSTAS)
+
+
+@compiled_test
+def test_simplest_mcxtrace():
+    instr = parse_mcstas_instr(dedent("""
+    define instrument test_simplest_mcxtrace(dummy=0.)
+    trace component origin = useless() at (0, 0, 0) absolute end
+    """), registries=[in_memory])
+    compile_and_run(instr, '-n 1 dummy=2', flavor=Flavor.MCXTRACE)
+

